### ISSUE TITLE
kube operator upgrade strategies

### ISSUE THREAD

**Issue by clux** [Babylonpartners/shipcat#65]
Writing down a couple of potential upgrade strategies for a tiller replacement. Using `raftcat` as an example kube operator even though it doesn't handle upgrades - but it still reads the CRDs.

## Scenario
shipcat upgrades change config/manifest syntax after the pin is bump so the operator can no longer parse the CRDs straight after they are applied.

## Solution 1: eventually consistent pins
Explicit upgrades of raftcat along with shipcat pins in config when bumped in manifests. Thus raftcat would only be slightly confused for a few minutes between the first CRDs being applied and itself being upgraded (or the other way around depending on the order chosen).

Less magical solution, but it would need to re-apply the world on re-upgrade because it lost the UPDATE events in the first minutes of downtime.

## Solution 2: eventually consistent self-upgrading operator
When raftcat syncs a config crd which is unparseable, parse it as untyped yaml and find the version field (present in config). Then tell kube to upgrade/downgrade raftcat (itself) to the version it found.

More magic solution, less pins, but would also need to re-apply the world after it boots because it lost a couple of UPDATE events (although less of them).

## Solution 3: ???
Maybe it's possible to pin versions and keep track of where to continue between upgrades. But don't see a clean way of doing that in an async world yet.

## Mitigating factors
We can make upgrades less breaking by only applying `#[serde(deny_unknown_fields)]` to all structs in the `filesystem` feature in `shipcat_definitions`. If we do that rigorously, and consider key renames / type changes as breaking changes to be done over several upgrades, then we can in theory avoid breaking changes.

This would be really hard to enforce though. Leaning towards handling it and maintaining strictness.

**Comment by clux** [comment]
[CRD versioning talk from kubecon](https://www.youtube.com/watch?v=HsYtMvvzDyI)

**Comment by janslow** [comment]
Another option is a [conversion webhook](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion). It's currently alpha, so we probably can't use it yet. However, it could be used to convert the current version of the CRD to the version supported by Raftcat.

This would also work well for other services (which can just read their supported version). Possibly using a `v1beta` version of the CRD as a stable CRD (and more minimal) and a `dev-0.107.0` or similar which is only used by Shipcat itself (so can have more breaking changes)

A simpler option is we ban breaking changes in a single version in favour of deprecation (maybe snapshot tests could help enforce this?).

**Comment by clux** [comment]
We have kind of implicitly settled on solution 1 in manifests, so we just need to write some logic for it. 

TL;DR: what we enforce:
every shipcat version must is accompanied by a raftcat bump (if raftcat is used).

This will ensure raftcat working for almost all of the time (save for maybe a minute during breaking schemas) because raftcat might crash once as new manifest schemas are applied, but that reconcile will also upgrade raftcat in the same go, so it fixes itself.

This feels like a pretty small price to pay for being able to make breaking changes to a schema.
